### ISSUE TITLE
Property or field 'duoAuthenticationService' cannot be found

### ISSUE THREAD

**Issue by jms18** [Unicon/shib-mfa-duo-auth#15]
Encountered this with IDP 3.2 and jetty 9.3 in conjunction with the shib-cas-authn3 plugin. Initially, I thought the problem might be with the Shibcas flow not properly passing off into sub-flow states; but the further I dug, the less likely that appeared. Traced the calling stack as best I could, and I began to think there might have been a problem with my classpath's, but the container was picking up everything but the method call.

I'm not sure if it is related to #12 or not. I'm going to keep stabbing at it and hope I can tease out more information. Attached is an abridged DEBUG trace of the idp and spring webflow (the full trace is too large to attach).
[shibb-idp-3.2.1-mfa-duo-auth-error-summary.txt](https://github.com/Unicon/shib-mfa-duo-auth/files/278497/shibb-idp-3.2.1-mfa-duo-auth-error-summary.txt)

**Comment by dhnukem** [comment]
I ran into the same issue on tomcat 8.  After much digging, I found that the annotations on the DuoAuthenticationService class were removed somewhere between 1.0.1 to 1.1.3:

```
-@Service('duoAuthenticationService')
 class DuoAuthenticationService {
-    @Value('%{duo.integrationKey}')
     private final String integrationKey

-    @Value('%{duo.secretKey}')
     private final String secretKey

-    @Value('%{duo.applicationKey}')
     private final String applicationKey

-    @Value('%{duo.apiHost}')
     private final String apiHost
```

I added the following bean to my global.xml file, and the error went away.

```
    <bean id="duoAuthenticationService" class="net.unicon.iam.shibboleth.idp.authn.duo.authn.DuoAuthenticationService"
        c:integrationKey="%{duo.integrationKey}"
        c:secretKey="%{duo.secretKey}"
        c:applicationKey="%{duo.applicationKey}"
        c:apiHost="%{duo.apiHost}"
        />
```

So, either the documentation should be updated, or maybe have this bean be incorporated into the default config in META-INF/shibboleth-idp/conf/global.xml?

**Comment by jms18** [comment]
That worked. Wish I would have thought of that. Many thanks.

**Comment by scalding** [comment]
Sorry for the late response on this. There's documentation out there on this somewhere, though, you're right that it should be here as well.

The reason for the removal of the annotations was to allow for easier configuration of multiple Duo services, for instance if you need to use Duo against two separate Duo integration configurations (say, one for students that might allow a pass through and another for employees that doesn't).

I do like the idea of defining the bean in the included `global.xml` file, which if you make the changes as noted to your `web.xml`, you should be able to easily override in one of the Shibboleth context files. I'll see what I can do about that soon.

**Comment by scalding** [comment]
This should be fixed in version 1.1.3. Thanks, and let me know if you run into any other problems.
